---
title: "`r paste('Gene-level mbGWAS,', taxLvl, 'level')`"
subtitle: "`r paste(genetic_dataset)`"
date: "`r Sys.Date()`"
output:
  rmdformats::robobook:
    highlight: kate
    toc_depth: 3
    number_sections: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}

## Global Options

opts <- options(
  knitr.kable.NA = "",
  knitr.duplicate.label = "allow")

knitr::opts_chunk$set(
  cache = FALSE,
  cache.path = "/scratch/gantonello/regenie_cache/",
  concordance = TRUE, 
  prompt = FALSE, # fig.width=5, fig.height=5,
  out.width = "100%",
  echo = FALSE, 
  warning = FALSE,
  message = FALSE,
  error = TRUE,
  #fig.keep = "none",
  tidy = FALSE,
  comment = ""
  )

library(tidyverse)
library(magrittr)
library(stringr)
library(forcats)
library(knitr)
library(DT)
library(ggrepel)
library(gridExtra)
library(data.table)
library(ggpubr)
library(kableExtra)
```

```{r paths setup, include=FALSE}
res.path <- file.path(base.path, "results", "results")
```

```{r, eval = interactive()}
library(optparse)
opts_args <- list(statistical_test = "ADD-SKATO-ACAT", threads = 8, p_value_threshold = "autoFind", verbose = TRUE, log10 = TRUE)
```

```{r}

regenie_gene_masks <- grep("regenie_gene_masks", basic_setup_gene_level, value = T) %>% 
  str_extract("\\'.*") %>% 
  gsub("'", "", .) %>% 
  readLines()

# format veps impact prediction #########
veps <- str_split(regenie_gene_masks, pattern = "\t", simplify = T) %>% .[,2] %>% 
  sapply(str_split, ",", simplify = T)

veps_impact_names <- str_split(regenie_gene_masks, pattern = "\t", simplify = T) %>% .[,1]

veps_impact <- lapply(1:length(veps), function(i) set_names(veps[[i]], rep(veps_impact_names[[i]], length(veps[[i]])))) %>% 
  unlist()

rm(veps)

##############################################################################

regenie_gene_annotation <- grep("regenie_gene_anno", basic_setup_gene_level, value = T) %>% 
  str_extract("\\'.*") %>%
  gsub("'", "", .) %>%
  fread(
    nThread = ceiling(opts_args$threads/4*3)
    ) %>% 
  set_names(c("ID_variant_full", "Gene", "VEP_annot")) %>% 
  mutate(
    GENPOS = str_split(ID_variant_full, ":", simplify = T) %>% .[, 2] %>% as.integer(),
    CHROM = str_extract(ID_variant_full, "^.*\\:[0-9]") %>% gsub("\\:[0-9]", "", .) %>% parse_number()
  ) %>% 
  left_join(data.frame(impact = names(veps_impact), VEP_annot = veps_impact), by = "VEP_annot") %>% 
  rename(GENE = Gene)


##########################################
#------------------
variants_per_gene <- regenie_gene_annotation %>% 
  filter(!is.na(impact)) %>%
  group_by(GENE, impact) %>% 
  tally() %>% 
  pivot_wider(names_from = "impact", values_from = "n") %>% 
  select(GENE, any_of(rev(veps_impact_names)))



variants_per_gene_filtered <- variants_per_gene[rowSums(is.na(variants_per_gene[-1])) < 3,]

if(opts_args$p_value_threshold=="autoFind"){
opts_args$p_value_threshold <- 0.05/nrow(variants_per_gene_filtered)*(1/readRDS(file.path(base.path, "..", "regenie_input", "n_components_0.95.rds")))
}
```

```{r}

files <- list.files(res.path, pattern = "regenie.gz$", full.names = T) 

if(opts_args$log10){
    opts_args$p_value_threshold <- round(-log10(opts_args$p_value_threshold),3)
    awk_command_readIn <- paste0(" | awk '{if ($12 > ", opts_args$p_value_threshold, ") {print}}'")
} else{
      awk_command_readIn <- paste0(" | awk '{if ($12 < ", opts_args$p_value_threshold, ") {print}}'")
}

system.time(
files_readIn <- lapply(files, function(x) fread(cmd = paste0("zcat ", x, awk_command_readIn), nThread = ceiling(opts_args$threads/2))) %>% 
  set_names(gsub(".regenie.gz", "", basename(files), fixed = T)) %>% 
  .[sapply(., nrow) > 0] %>% 
  bind_rows(.id = "TAXON") %>% 
  rename(MASK = ALLELE1)
)

masks_names_unique <- paste(paste0("\\.", unique(files_readIn$MASK)), collapse = "|")
  
input_significant_hits  <- files_readIn %>% 
  mutate(GENE = gsub(masks_names_unique, "", ID)) %>% 
  select(-ID) %>% 
  relocate(GENE, .after = "GENPOS") %>% 
  left_join(taxonomy_info %>% rename(TAXON = taxLvl) %>% select(any_of(c("TAXON", "Species", "Phylum"))), by = "TAXON") %>% 
  filter(TEST == opts_args$statistical_test)

source("https://raw.githubusercontent.com/g-antonello/mbGWAS_regenie/main/customFunctions.GA.R")
datatable(input_significant_hits %>% 
            mutate(GENE = link_GeneCards(GENE)), 
          caption = paste0("list of significant associations with -log10(P-value) >", format.pval(opts_args$p_value_threshold, digits = 1)),
          escape = FALSE)
```

# Number of variants tested per locus (gene)

```{r}
datatable(variants_per_gene_filtered, caption = "Number of variants for each impact predicted by VEP. These are the variants tested with Gene-level analysis, NOT the results")
```

# Gene level analysis: significant genes

```{r}

input_significant_hits %>% 
  group_by(TAXON, GENE, MASK) %>%
  summarise(LOG10P = round(LOG10P, 2)) %>% 
  pivot_wider(names_from = MASK, values_from = LOG10P) %>% 
  mutate_at(vars(all_of(contains("impact"))), .funs = function(x) ifelse(is.na(x), "", x)) %>% 
  datatable(caption = paste("P-values of the association for each ASV-gene-mask combination. Empty fields are non-significant associations (P-value below the chosen threshold:", opts_args$p_value_threshold))

```



